clients/drcachesim/tracer/output.cpp

```diff
@@ -752,8 +751,23 @@ output_buffer(void *drcontext, per_thread_t *data, byte *buf_base, byte *buf_ptr
                     DR_ASSERT(is_ok_to_split_before(
                         instru->get_entry_type(pipe_start + header_size),
                         instru->get_entry_size(pipe_start + header_size)));
-                    pipe_start = atomic_pipe_write(drcontext, pipe_start, pipe_end,
-                                                   get_local_window(data));
+                    // Check if we went over the edge waiting for enough entries to
+                    // write. If we did, we simply write till the last ok-to-split ref.
```

Comment:
Is there a reason this kept happening only with scatter/gather?  Do we understand why?  If so, can we use that knowledge to make a targeted test that always reproduces the issue?

Reply:
I think it's because there's an underlying assumption here that we'll re-do the check at L749 sooner than every `(MAX_NUM_DELAY_ENTRIES +1)` (==5) entries, but with scatter-gather I'm seeing 8 intervening memrefs. So we go over the threshold by the time we get the opportunity to check whether we should output data already.

Reply:
The existing scatter-gather test seems targeted enough? It's hitting this case consistently.

Reply:
Added more details to description